app/slave/cluster_slave.py

```diff
@@ -56,6 +58,41 @@ def __init__(self, port, host, num_executors=10):
         self._build_teardown_coin = None
         self._base_executor_index = None
 
+        # Configure heartbeat
+        self._heartbeat_failure_count = 0
+        self._heartbeat_failure_threshold = Configuration['heartbeat_failure_threshold']
+        self._heartbeat_interval = Configuration['heartbeat_interval']
+        self._hb_scheduler = sched.scheduler()
+
+    def start_heartbeat_thread(self):
+        self._logger.info('Heartbeat will run every {} seconds'.format(self._heartbeat_interval))
+        SafeThread(target=self._start_heartbeat, name='HeartbeatThread', daemon=True).start()
+
+    def _start_heartbeat(self):
+        self._hb_scheduler.enter(0, 0, self._run_heartbeat)
+        self._hb_scheduler.run()
+
+    def _run_heartbeat(self):
+        try:
+            self._send_heartbeat_to_master()
+            self._heartbeat_failure_count = 0
+        except (requests.ConnectionError, requests.Timeout):
+            self._heartbeat_failure_count += 1
+            if self._heartbeat_failure_count >= self._heartbeat_failure_threshold:
+                self._logger.error('Master is not responding to heartbeats')
+
+                # TODO: Right now the slave simply dies when it does not hear back from master. The next step would
+                # be to try to reconnect to master at this point. In future the heartbeat and connect_to_master
+                # methods can combined into one. This combined method will behave differently based on current state.
+                self.kill()
```

Comment:
Why even kill it? Why not just let it stay up forever? Indefinitely spamming the master with requests.

I suppose this is gonna change really soon anyway, so nbd.

Reply:
I was just talking about the comment @mayurva left

```TODO: Right now the slave simply dies when it does not hear back from master.
The next step would be to try to reconnect to master at this point. In future
the heartbeat and connect_to_master methods can combined into one. This
combined method will behave differently based on current state.```

I think the plan is for this to happen soon